pkg/reconciler/bitbucketsource/resources/bitbucket_client.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/knative/pkg/logging"
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const (
+	defaultBaseUrl   = "https://api.bitbucket.org/2.0/"
+	defaultUserAgent = "go-bitbucket"
+	mediaTypeJson    = "application/json"
+)
+
+// Hook struct to marshal/unmarshal BitBucket requests/responses.
+type Hook struct {
+	URL         string   `json:"url,omitempty"`
+	Description string   `json:"description,omitempty"`
+	Events      []string `json:"events,omitempty"`
+	Active      bool     `json:"active,omitempty"`
+	UUID        string   `json:"uuid,omitempty"`
+}
+
+// Client struct used to send http.Requests to BitBucket.
+type Client struct {
+	client    *http.Client
+	logger    *zap.SugaredLogger
+	token     *oauth2.Token
+	baseUrl   *url.URL
+	userAgent string
+}
+
+// NewClient creates a new Client for sending http.Requests to BitBucket.
+func NewClient(ctx context.Context, token *oauth2.Token) *Client {
+	logger := logging.FromContext(ctx)
+	httpClient := http.DefaultClient
+	baseUrl, _ := url.Parse(defaultBaseUrl)
+	return &Client{client: httpClient, logger: logger, token: token, baseUrl: baseUrl, userAgent: defaultUserAgent}
+}
+
+// CreateHook creates a WebHook for 'owner' and 'repo'.
+func (c *Client) CreateHook(owner, repo string, hook *Hook) (*Hook, error) {
+	body, err := createHookBody(hook)
+	if err != nil {
+		return nil, err
+	}
+	var urlStr string
+	if repo == "" {
+		// For every repo of the owner.
+		urlStr = fmt.Sprintf("teams/%s/hooks", owner)
+	} else {
+		// For a specific repo of the owner.
+		urlStr = fmt.Sprintf("repositories/%s/%s/hooks", owner, repo)
+	}
+
+	h := new(Hook)
+	err = c.doRequest("POST", urlStr, body, h)
+	return h, err
+}
+
+// DeleteHook deletes the WebHook 'hookUUID' previously registered for 'owner' and 'repo'.
+func (c *Client) DeleteHook(owner, repo, hookUUID string) error {
+
+	var urlStr string
+	if repo == "" {
+		urlStr = fmt.Sprintf("teams/%v/hooks/%s", owner, hookUUID)
+	} else {
+		urlStr = fmt.Sprintf("repositories/%v/%v/hooks/%s", owner, repo, hookUUID)
```

Comment:
`%v` -> `%s` because `owner` and `repo` are strings

Reply:
Done